zetaclient/chains/solana/observer/outbound.go

```diff
@@ -2,22 +2,344 @@ package observer
+	"github.com/gagliardetto/solana-go"
+	"github.com/gagliardetto/solana-go/rpc"
+	"github.com/near/borsh-go"
+	"github.com/pkg/errors"
 	"github.com/rs/zerolog"
 
-	"github.com/zeta-chain/zetacore/x/crosschain/types"
+	"github.com/zeta-chain/zetacore/pkg/chains"
+	"github.com/zeta-chain/zetacore/pkg/coin"
+	contract "github.com/zeta-chain/zetacore/pkg/contract/solana"
+	crosschaintypes "github.com/zeta-chain/zetacore/x/crosschain/types"
+	"github.com/zeta-chain/zetacore/zetaclient/chains/interfaces"
+	zctx "github.com/zeta-chain/zetacore/zetaclient/context"
+	clienttypes "github.com/zeta-chain/zetacore/zetaclient/types"
+	"github.com/zeta-chain/zetacore/zetaclient/zetacore"
 )
 
 // GetTxID returns a unique id for Solana outbound
-func (ob *Observer) GetTxID(_ uint64) string {
-	return ""
+func (ob *Observer) GetTxID(nonce uint64) string {
+	tssAddr := ob.TSS().EVMAddress().String()
+	return fmt.Sprintf("%d-%s-%d", ob.Chain().ChainId, tssAddr, nonce)
 }
 
-// IsOutboundProcessed checks outbound status and returns (isIncluded, isConfirmed, error)
+// WatchOutbound watches evm chain for outgoing txs status
+// TODO(revamp): move ticker function to ticker file
+// TODO(revamp): move inner logic to a separate function
+func (ob *Observer) WatchOutbound(ctx context.Context) error {
+	// get app context
+	app, err := zctx.FromContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	// create outbound ticker based on chain params
+	chainID := ob.Chain().ChainId
+	ticker, err := clienttypes.NewDynamicTicker(
+		fmt.Sprintf("Solana_WatchOutbound_%d", chainID),
+		ob.GetChainParams().OutboundTicker,
+	)
+	if err != nil {
+		ob.Logger().Outbound.Error().Err(err).Msg("error creating ticker")
+		return err
+	}
+
+	ob.Logger().Outbound.Info().Msgf("WatchOutbound started for chain %d", chainID)
+	sampledLogger := ob.Logger().Outbound.Sample(&zerolog.BasicSampler{N: 10})
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C():
+			if !app.IsOutboundObservationEnabled(ob.GetChainParams()) {
+				sampledLogger.Info().Msgf("WatchOutbound: outbound observation is disabled for chain %d", chainID)
+				continue
+			}
+			trackers, err := ob.ZetacoreClient().GetAllOutboundTrackerByChain(ctx, chainID, interfaces.Ascending)
+			if err != nil {
+				continue
+			}
+
+			for _, tracker := range trackers {
+				// go to next tracker if this one already has a finalized tx
+				nonce := tracker.Nonce
+				if ob.IsTxFinalized(tracker.Nonce) {
+					continue
+				}
+
+				// get original cctx parameters
+				cctx, err := ob.ZetacoreClient().GetCctxByNonce(ctx, chainID, tracker.Nonce)
+				if err != nil {
+					// take a rest if zetacore RPC breaks
+					ob.Logger().Outbound.Error().
+						Err(err).
+						Msgf("WatchOutbound: can't find cctx for chain %d nonce %d", chainID, tracker.Nonce)
+					break
+				}
+				coinType := cctx.InboundParams.CoinType
+
+				txCount := 0
+				var txResult *rpc.GetTransactionResult
+				for _, txHash := range tracker.HashList {
```

Comment:
Can we explain in comment what do we do with the tx hash?
The function is called `WatchOutbound` but it is unclear from the function body what we actually do with the outbounds.

Reply:
additional comment added https://github.com/zeta-chain/node/pull/2560/commits/fc0c9bf58179a0da2c07b996c983c62c1aad901d